File: test/extensions/language/alignment/java/alignment/AutoOptionalMemberAlignmentTest.java
package alignment;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

import alignment.auto_optional_member_alignment.AutoOptionalMemberAlignment;

public class AutoOptionalMemberAlignmentTest
{
    @Test
    public void readWithOptional() throws IOException, ZserioError
    {
        final int autoOptionalField = 0x1234;
        final int field = 0x7654;
        final BitBuffer buffer = writeAutoOptionalMemberAlignmentToBitBuffer(autoOptionalField, field);
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                SerializeUtil.deserialize(AutoOptionalMemberAlignment.class, buffer);
        checkAutoOptionalMemberAlignment(autoOptionalMemberAlignment, autoOptionalField, field);
    }

    @Test
    public void readWithoutOptional() throws IOException, ZserioError
    {
        final int field = 0x2222;
        final BitBuffer buffer = writeAutoOptionalMemberAlignmentToBitBuffer(null, field);
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                SerializeUtil.deserialize(AutoOptionalMemberAlignment.class, buffer);
        checkAutoOptionalMemberAlignment(autoOptionalMemberAlignment, null, field);
    }

    @Test
    public void bitSizeOfWithOptional()
    {
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(0x4433, 0x1122);
        assertEquals(WITH_AUTO_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE, autoOptionalMemberAlignment.bitSizeOf());
    }

    @Test
    public void bitSizeOfWithoutOptional()
    {
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(null, 0x7624);
        assertEquals(WITHOUT_AUTO_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE, autoOptionalMemberAlignment.bitSizeOf());
    }

    @Test
    public void initializeOffsetsWithOptional()
    {
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(0x1111, 0x3333);
        int bitPosition = 0;
        for (; bitPosition < 32; ++bitPosition)
            assertEquals(WITH_AUTO_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE,
                    autoOptionalMemberAlignment.initializeOffsets(bitPosition));

        assertEquals(WITH_AUTO_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE + bitPosition,
                autoOptionalMemberAlignment.initializeOffsets(bitPosition));
    }

    @Test
    public void initializeOffsetsWithoutOptional()
    {
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(null, 0x3334);
        final int bitPosition = 1;
        assertEquals(WITHOUT_AUTO_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE + bitPosition,
                autoOptionalMemberAlignment.initializeOffsets(bitPosition));
    }

    @Test
    public void writeWithOptional() throws IOException, ZserioError
    {
        final int autoOptionalField = 0x9ADB;
        final int field = 0x8ACD;
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(autoOptionalField, field);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        autoOptionalMemberAlignment.write(writer);

        final ByteArrayBitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final AutoOptionalMemberAlignment readAutoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(reader);
        checkAutoOptionalMemberAlignment(readAutoOptionalMemberAlignment, autoOptionalField, field);
        assertTrue(autoOptionalMemberAlignment.equals(readAutoOptionalMemberAlignment));
    }

    @Test
    public void writeWithoutOptional() throws IOException, ZserioError
    {
        final int field = 0x7ACF;
        final AutoOptionalMemberAlignment autoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(null, field);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        autoOptionalMemberAlignment.write(writer);

        final ByteArrayBitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final AutoOptionalMemberAlignment readAutoOptionalMemberAlignment =
                new AutoOptionalMemberAlignment(reader);
        checkAutoOptionalMemberAlignment(readAutoOptionalMemberAlignment, null, field);
        assertTrue(autoOptionalMemberAlignment.equals(readAutoOptionalMemberAlignment));
    }

    private BitBuffer writeAutoOptionalMemberAlignmentToBitBuffer(Integer autoOptionalField, int field)
            throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            if (autoOptionalField != null)
            {
                writer.writeBool(true);
                writer.writeBits(0, 31);
                writer.writeInt(autoOptionalField);
            }
            else
            {
                writer.writeBool(false);
            }
            writer.writeInt(field);

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private void checkAutoOptionalMemberAlignment(
            AutoOptionalMemberAlignment autoOptionalMemberAlignment, Integer autoOptionalField, int field)
    {
        if (autoOptionalField != null)
        {
            assertEquals(autoOptionalField, autoOptionalMemberAlignment.getAutoOptionalField());
            assertTrue(autoOptionalMemberAlignment.isAutoOptionalFieldUsed());
        }
        else
        {
            assertFalse(autoOptionalMemberAlignment.isAutoOptionalFieldUsed());
        }
        assertEquals(field, (int)autoOptionalMemberAlignment.getField());
    }

    private static final int WITH_AUTO_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 96;
    private static final int WITHOUT_AUTO_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 33;
}

File: test/extensions/language/alignment/cpp/AutoOptionalMemberAlignmentTest.cpp
#include "alignment/auto_optional_member_alignment/AutoOptionalMemberAlignment.h"
#include "gtest/gtest.h"
#include "zserio/BitStreamReader.h"
#include "zserio/BitStreamWriter.h"

namespace alignment
{
namespace auto_optional_member_alignment
{

class AutoOptionalMemberAlignmentTest : public ::testing::Test
{
protected:
    void writeAutoOptionalMemberAlignmentToByteArray(
            zserio::BitStreamWriter& writer, bool hasAutoOptional, int32_t autoOptionalField, int32_t field)
    {
        writer.writeBool(hasAutoOptional);

        if (hasAutoOptional)
        {
            writer.writeBits(0, 31);
            writer.writeSignedBits(autoOptionalField, 32);
        }

        writer.writeSignedBits(field, 32);
    }

    void checkAutoOptionalMemberAlignment(const AutoOptionalMemberAlignment& autoOptionalMemberAlignment,
            bool hasAutoOptional, int32_t autoOptionalField, int32_t field)
    {
        if (hasAutoOptional)
        {
            ASSERT_EQ(autoOptionalField, autoOptionalMemberAlignment.getAutoOptionalField());
            ASSERT_TRUE(autoOptionalMemberAlignment.isAutoOptionalFieldUsed());
        }
        else
        {
            ASSERT_FALSE(autoOptionalMemberAlignment.isAutoOptionalFieldUsed());
        }

        ASSERT_EQ(field, autoOptionalMemberAlignment.getField());
    }

    void fillAutoOptionalMemberAlignment(AutoOptionalMemberAlignment& autoOptionalMemberAlignment,
            bool hasAutoOptional, int32_t autoOptionalField, int32_t field)
    {
        if (hasAutoOptional)
        {
            autoOptionalMemberAlignment.setAutoOptionalField(autoOptionalField);
        }
        autoOptionalMemberAlignment.setField(field);
    }

    static const size_t WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    static const size_t WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const size_t AutoOptionalMemberAlignmentTest::WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 96;
const size_t AutoOptionalMemberAlignmentTest::WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE = 33;

TEST_F(AutoOptionalMemberAlignmentTest, readWithOptional)
{
    const bool hasAutoOptional = true;
    const int32_t autoOptionalField = 0x1234;
    const int32_t field = 0x7654;
    zserio::BitStreamWriter writer(bitBuffer);
    writeAutoOptionalMemberAlignmentToByteArray(writer, hasAutoOptional, autoOptionalField, field);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    AutoOptionalMemberAlignment autoOptionalMemberAlignment(reader);
    checkAutoOptionalMemberAlignment(autoOptionalMemberAlignment, hasAutoOptional, autoOptionalField, field);
}

TEST_F(AutoOptionalMemberAlignmentTest, readWithoutOptional)
{
    const bool hasAutoOptional = false;
    const int32_t autoOptionalField = 0;
    const int32_t field = 0x2222;
    zserio::BitStreamWriter writer(bitBuffer);
    writeAutoOptionalMemberAlignmentToByteArray(writer, hasAutoOptional, autoOptionalField, field);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    AutoOptionalMemberAlignment autoOptionalMemberAlignment(reader);
    checkAutoOptionalMemberAlignment(autoOptionalMemberAlignment, hasAutoOptional, autoOptionalField, field);
}

TEST_F(AutoOptionalMemberAlignmentTest, bitSizeOfWithOptional)
{
    AutoOptionalMemberAlignment autoOptionalMemberAlignment;
    fillAutoOptionalMemberAlignment(autoOptionalMemberAlignment, true, 0x4433, 0x1122);

    const size_t expectedBitSize = WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, autoOptionalMemberAlignment.bitSizeOf());
}

TEST_F(AutoOptionalMemberAlignmentTest, bitSizeOfWithoutOptional)
{
    AutoOptionalMemberAlignment autoOptionalMemberAlignment;
    fillAutoOptionalMemberAlignment(autoOptionalMemberAlignment, false, 0, 0x7624);

    const size_t expectedBitSize = WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, autoOptionalMemberAlignment.bitSizeOf());
}

TEST_F(AutoOptionalMemberAlignmentTest, initializeOffsetsWithOptional)
{
    AutoOptionalMemberAlignment autoOptionalMemberAlignment;
    fillAutoOptionalMemberAlignment(autoOptionalMemberAlignment, true, 0x1111, 0x3333);

    const size_t expectedBitSize = WITH_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    size_t bitPosition = 0;
    for (; bitPosition < 32; ++bitPosition)
    {
        ASSERT_EQ(expectedBitSize, autoOptionalMemberAlignment.initializeOffsets(bitPosition));
    }
    ASSERT_EQ(expectedBitSize + bitPosition, autoOptionalMemberAlignment.initializeOffsets(bitPosition));
}

TEST_F(AutoOptionalMemberAlignmentTest, initializeOffsetsWithoutOptional)
{
    AutoOptionalMemberAlignment autoOptionalMemberAlignment;
    fillAutoOptionalMemberAlignment(autoOptionalMemberAlignment, false, 0, 0x3334);

    const size_t expectedBitSize = WITHOUT_OPTIONAL_MEMBER_ALIGNMENT_BIT_SIZE;
    const size_t bitPosition = 1;
    ASSERT_EQ(expectedBitSize + bitPosition, autoOptionalMemberAlignment.initializeOffsets(bitPosition));
}

TEST_F(AutoOptionalMemberAlignmentTest, writeWithOptional)
{
    const bool hasAutoOptional = true;
    const int32_t autoOptionalField = 0x9ADB;
    const int32_t field = 0x8ACD;
    AutoOptionalMemberAlignment autoOptionalMemberAlignment;
    fillAutoOptionalMemberAlignment(autoOptionalMemberAlignment, hasAutoOptional, autoOptionalField, field);

    zserio::BitStreamWriter writer(bitBuffer);
    autoOptionalMemberAlignment.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    AutoOptionalMemberAlignment readAutoOptionalMemberAlignment(reader);
    checkAutoOptionalMemberAlignment(
            readAutoOptionalMemberAlignment, hasAutoOptional, autoOptionalField, field);
    ASSERT_TRUE(autoOptionalMemberAlignment == readAutoOptionalMemberAlignment);
}

TEST_F(AutoOptionalMemberAlignmentTest, writeWithoutOptional)
{
    const bool hasAutoOptional = true;
    const int32_t autoOptionalField = 0;
    const int32_t field = 0x7ACF;
    AutoOptionalMemberAlignment autoOptionalMemberAlignment;
    fillAutoOptionalMemberAlignment(autoOptionalMemberAlignment, hasAutoOptional, autoOptionalField, field);

    zserio::BitStreamWriter writer(bitBuffer);
    autoOptionalMemberAlignment.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    AutoOptionalMemberAlignment readAutoOptionalMemberAlignment(reader);
    checkAutoOptionalMemberAlignment(
            readAutoOptionalMemberAlignment, hasAutoOptional, autoOptionalField, field);
    ASSERT_TRUE(autoOptionalMemberAlignment == readAutoOptionalMemberAlignment);
}

} // namespace auto_optional_member_alignment
} // namespace alignment
